File: frontend/projects/athlete/src/app/data/tournament-announcements-repository.ts
import { collection, getDocs, limit, orderBy, query, where, type Firestore } from 'firebase/firestore';

/** Avisos que o organizador publica durante o torneio. Vivem na MESMA coleção do feed da
 *  comunidade (`communityFeed`), gravados por `postTournamentAnnouncement`
 *  (functions/src/tournament-announcements.ts) com `type: 'organizer_announcement'` — por isso
 *  não têm coleção própria. `watchCommunityFeed` descarta esse tipo de propósito (o feed da
 *  Comunidade só mostra abertura de inscrição e campeões), então a leitura por torneio fica aqui.
 *
 *  Exige o índice composto `type ASC, tournamentId ASC, createdAt DESC` (firestore.indexes.json). */

export interface TournamentAnnouncement {
  id: string;
  message: string;
  createdAt: Date | null;
}

function str(v: unknown): string | null {
  return typeof v === 'string' && v.trim() ? v.trim() : null;
}

function toDate(v: unknown): Date | null {
  const t = v as { toDate?: () => Date } | undefined;
  return typeof t?.toDate === 'function' ? t.toDate() : null;
}

export async function fetchTournamentAnnouncements(db: Firestore, tournamentId: string, max = 10): Promise<TournamentAnnouncement[]> {
  if (!tournamentId) return [];
  const snap = await getDocs(
    query(
      collection(db, 'communityFeed'),
      where('type', '==', 'organizer_announcement'),
      where('tournamentId', '==', tournamentId),
      orderBy('createdAt', 'desc'),
      limit(max),
    ),
  );
  return snap.docs
    .map((d) => {
      const data = d.data() as Record<string, unknown>;
      const message = str(data['message']);
      return message ? { id: d.id, message, createdAt: toDate(data['createdAt']) } : null;
    })
    .filter((a): a is TournamentAnnouncement => a != null);
}

File: nexago_app/lib/features/tournaments/data/tournament_announcements_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// Um aviso que o organizador publicou durante o torneio.
class TournamentAnnouncement {
  const TournamentAnnouncement({
    required this.id,
    required this.message,
    required this.createdAt,
  });

  final String id;
  final String message;
  final DateTime? createdAt;
}

/// Avisos do organizador de um torneio.
///
/// Vivem na MESMA coleção do feed da comunidade (`communityFeed`), gravados com
/// `type: 'organizer_announcement'` — não têm coleção própria. O feed da
/// Comunidade descarta esse tipo de propósito (lá só entram abertura de
/// inscrição e campeões), então a leitura por torneio fica aqui.
///
/// Exige o índice composto `type ASC, tournamentId ASC, createdAt DESC`, que já
/// está em `firestore.indexes.json`.
class TournamentAnnouncementsRepository {
  const TournamentAnnouncementsRepository(this._db);

  final FirebaseFirestore _db;

  static const int _limit = 10;

  Stream<List<TournamentAnnouncement>> watchByTournament(String tournamentId) {
    final id = tournamentId.trim();
    if (id.isEmpty) return Stream.value(const []);

    return _db
        .collection('communityFeed')
        .where('type', isEqualTo: 'organizer_announcement')
        .where('tournamentId', isEqualTo: id)
        .orderBy('createdAt', descending: true)
        .limit(_limit)
        .snapshots()
        .map((snap) => snap.docs.map(_fromDoc).toList());
  }

  static TournamentAnnouncement _fromDoc(
    QueryDocumentSnapshot<Map<String, dynamic>> doc,
  ) {
    final data = doc.data();
    final raw = data['createdAt'];
    return TournamentAnnouncement(
      id: doc.id,
      message: (data['message'] as String?)?.trim() ??
          (data['body'] as String?)?.trim() ??
          '',
      createdAt: raw is Timestamp ? raw.toDate() : null,
    );
  }
}

/// Falha degrada para lista vazia: um aviso que não carrega não pode derrubar a
/// seção "Agora", que é a tela que o atleta olha no dia de jogo.
final tournamentAnnouncementsProvider = StreamProvider.autoDispose
    .family<List<TournamentAnnouncement>, String>((ref, tournamentId) {
  return TournamentAnnouncementsRepository(FirebaseFirestore.instance)
      .watchByTournament(tournamentId)
      .handleError((_) => const <TournamentAnnouncement>[]);
});
